compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -766,8 +775,27 @@ fn may_define_opaque_type<'tcx>(
     def_id: LocalDefId,
     opaque_def_id: LocalDefId,
     span: Span,
-) -> Result<(), LocalDefId> {
-    if tcx.is_descendant_of(opaque_def_id.to_def_id(), def_id.to_def_id()) {
+) -> Result<(), (LocalDefId, bool)> {
+    let mut parent = tcx.local_parent(opaque_def_id);
+    loop {
+        trace!(?parent);
+        match tcx.def_kind(parent) {
+            DefKind::AssocTy | DefKind::TyAlias => {
+                parent = tcx.local_parent(parent);
+                break;
+            }
+            // Skip nested opaque types
+            DefKind::OpaqueTy => {
+                parent = tcx.local_parent(parent);
+            }
+            def_kind => {
+                trace!(?def_kind);
+                return Err((def_id, false));
```

Comment:
please add a comment to this method explaining the meaning of the bool, or probably better, make the error case a struct